pkg/virtual/apiexport/authorizer/maximal_permission_policy.go

```diff
@@ -0,0 +1,176 @@
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clusters"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/authorization/delegated"
+	apisinformers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/indexers"
+	dynamiccontext "github.com/kcp-dev/kcp/pkg/virtual/framework/dynamic/context"
+)
+
+const (
+	VirtualAPIExportMaximumPermissionPolicyAuditPrefix   = "virtual.apiexport.maxpermissionpolicy.authorization.kcp.dev/"
+	VirtualAPIExportMaximumPermissionPolicyAuditDecision = VirtualAPIExportMaximumPermissionPolicyAuditPrefix + "decision"
+	VirtualAPIExportMaximumPermissionPolicyAuditReason   = VirtualAPIExportMaximumPermissionPolicyAuditPrefix + "reason"
+)
+
+type maximalPermissionAuthorizer struct {
+	getAPIExport            func(clusterName, apiExportName string) (*apisv1alpha1.APIExport, error)
+	newDeepSARAuthorizer    func(clusterName logicalcluster.Name) (authorizer.Authorizer, error)
+	getAPIExportsByIdentity func(identityHash string) ([]*apisv1alpha1.APIExport, error)
+}
+
+// NewMaximalPermissionAuthorizer creates an authorizer that checks the maximal permission policy
+// for the requested resource if the resource is a claimed resource in the requested API export.
+// The check is omitted if the requested resource itself is not associated with another API export.
+//
+// If the request is a cluster request the authorizer skips authorization if the request is not for a bound resource.
+// If the request is a wildcard request this check is skipped because no unique API binding can be determined.
+func NewMaximalPermissionAuthorizer(deepSARClient kubernetes.ClusterInterface, apiExportInformer apisinformers.APIExportInformer, apiBindingInformer apisinformers.APIBindingInformer) authorizer.Authorizer {
+	apiExportLister := apiExportInformer.Lister()
+	apiExportIndexer := apiExportInformer.Informer().GetIndexer()
+
+	return &maximalPermissionAuthorizer{
+		getAPIExport: func(clusterName, apiExportName string) (*apisv1alpha1.APIExport, error) {
+			return apiExportLister.Get(clusters.ToClusterAwareKey(logicalcluster.New(clusterName), apiExportName))
+		},
+		getAPIExportsByIdentity: func(identityHash string) ([]*apisv1alpha1.APIExport, error) {
+			return indexers.ByIndex[*apisv1alpha1.APIExport](apiExportIndexer, indexers.APIExportByIdentity, identityHash)
+		},
+		newDeepSARAuthorizer: func(clusterName logicalcluster.Name) (authorizer.Authorizer, error) {
+			return delegated.NewDelegatedAuthorizer(clusterName, deepSARClient)
+		},
+	}
+}
+
+func (a *maximalPermissionAuthorizer) Authorize(ctx context.Context, attr authorizer.Attributes) (authorizer.Decision, string, error) {
+	apiDomainKey := dynamiccontext.APIDomainKeyFrom(ctx)
+	parts := strings.Split(string(apiDomainKey), "/")
+	if len(parts) < 2 {
+		return authorizer.DecisionNoOpinion, VirtualAPIExportNotPermittedReason, fmt.Errorf("invalid API domain key")
```

Comment:
Are those user-facing errors?

Reply:
I just checked, yes these bubble down to the user.

Reply:
yes, indeed, and i am realizing most authorizers bubble down errors as well. I need to wrap this with a anonymizer+audit logger 🤔